File: leetcode/math/634_Find_the_Derangement_of_An_Array/codes/dp.py
MOD = 10**9 + 7

class Solution:
    def findDerangement(self, n: int) -> int:
        if n == 1:
            return 0
        if n == 2:
            return 1
        dp = [0] * (n + 1)
        dp[1] = 0
        dp[2] = 1
        for i in range(3, n + 1):
            dp[i] = ((i - 1) * (dp[i - 1] + dp[i - 2])) % MOD
        return dp[n]

File: leetcode/math/634_Find_the_Derangement_of_An_Array/codes/dp.c
#include <stdio.h>
#include <stdlib.h>

#define MOD 1000000007

int findDerangement(int n) {
    if (n == 1) return 0;
    if (n == 2) return 1;

    long long* dp = (long long*)malloc((n + 1) * sizeof(long long));
    dp[1] = 0;
    dp[2] = 1;

    for (int i = 3; i <= n; i++) {
        dp[i] = ((i - 1LL) * (dp[i - 1] + dp[i - 2])) % MOD;
    }

    int result = dp[n];
    free(dp);
    return result;
}
